Python/Module6_Testing/Pytest.md

```diff
@@ -0,0 +1,750 @@
+There is a project, "Nose2", which is carrying the torch of `nose`. However, this is a fledgling project by comparison to `pytest`.
+As of writing this, `pytest` was downloaded 12 million times last month versus `nose2`'s 150 thousand downloads.
+    
+The takeaway here is that, when it comes to picking a testing framework for Python, `pytest` is the clear choice.
+Any discussion that you come across to the contrary is likely outdated.
+</div>
+
+<!-- #region -->
+## Creating a Python Package with Tests
+
+It's time to create a proper test suite.
+Before proceeding any further, we should reread the material presented in [Module 5 - Import: Modules and Packages](https://www.pythonlikeyoumeanit.com/Module5_OddsAndEnds/Modules_and_Packages.html) and recall the essentials of import statements, modules, and Python packages.
+This material serves as the foundation for this section.
+
+### Organizing our Source Code
+Let's create the a Python package, which we will call `plymi_mod6`, with the following directory structure:
+
+```
+project_dir/     # the "parent directory" houses our source code, tests, and all other relevant files
+  - setup.py     # script responsible for installing `plymi_mod6` package
+  - plymi_mod6/  # directory containing source code of `plymi_mod6` package
+      |-- __init__.py
+      |-- basic_functions.py
+      |-- numpy_functions.py
+  - tests/        # test-suite for `plymi_mod6` package (to be run using pytest)
+      |-- conftest.py # optional configuration file for pytest
+      |-- test_basic_functions.py
+      |-- test_numpy_functions.py
+```
+
+A reference implementation of this package can be found [in this GitHub repository](https://github.com/rsokl/plymi_mod6).
+Populate the `basic_functions.py` file with the two functions that we were using as our source code in the previous section: `count_vowels` and `merge_max_mappings`.
+In in the `numpy_functions.py` module, add the `pairwise_dists` function that appears in [Module 3's discussion of optimized pairwise distances](https://www.pythonlikeyoumeanit.com/Module3_IntroducingNumpy/Broadcasting.html#Optimized-Pairwise-Distances).
+Don't forget to include `import numpy as np` in your script in accordance with how `pairwise_dists` calls NumPy functions. 
+
+We have arranged these functions so that they can be imported from the `basic_functions` module and the `numpy_functions` module, respectively, which reside in our `plymi_mod6` package.
+Let's fill out our `setup.py` script and install this package so that we can import it regardless of our current working directory. The content of `setup.py` will be:
+
+```python
+from setuptools import find_packages, setup
+
+setup(
+    name="plymi_mod6",
+    packages=find_packages(exclude=["tests", "tests.*"]),
+    version="1.0.0",
+    author="Your Name",
+    description="A template Python package for learning about testing",
+    install_requires=["numpy >= 1.10.0"],
+    tests_require=["pytest", "hypothesis"],
```

Comment:
Go ahead and `tests_require=["pytest>=5.3", "hypothesis>=5.0"]` - people are going to copy-and-paste this, and so we might as well pin them to modern versions by default :wink: (and `numpy>=1.17` while you're at it)